stl/inc/ranges

```diff
@@ -636,7 +634,87 @@ namespace ranges {
         /* [[no_unique_address]] */ _Ty _Value{};
     };
 
-    // CLASS TEMPLATE ranges::empty_view
+    template <_Destructible_object _Ty>
+    class _Non_propagating_cache { // a simplified optional that resets on copy / move
+    public:
+        constexpr _Non_propagating_cache() noexcept {}
+
+        constexpr ~_Non_propagating_cache() {
+            if (_Engaged) {
+                _Val.~_Ty();
+            }
+        }
+
+        // clang-format off
+        ~_Non_propagating_cache() requires is_trivially_destructible_v<_Ty> = default;
+        // clang-format on
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept {}
+
+        constexpr _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Engaged = false;
+            }
+        }
+
+        constexpr _Non_propagating_cache& operator=(const _Non_propagating_cache& _Other) noexcept {
+            if (_STD addressof(_Other) == this) {
+                return *this;
+            }
+
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            return *this;
+        }
+
+        constexpr _Non_propagating_cache& operator=(_Non_propagating_cache&& _Other) noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        template <class _It>
+        constexpr _Ty& _Emplace_deref(_It& _Iter) noexcept(
+            is_nothrow_constructible_v<_Ty, iter_reference_t<_It>>) /* strengthened */ {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            _Construct_in_place(_Val, *_Iter);
+            _Engaged = true;
+
+            return _Val;
+        }

```

Comment:
Just reiterating my comment that this is not to spec: _`emplace-deref`_ prohibits implementations from materializing `*i` before it is used to initialize the contained object (in other words, it requires guaranteed copy elision), and `_Construct_in_place(_Val, *_Iter);` fails that requirement because `*_Iter` is materialized when it is bound to the reference parameter of `_Construct_in_place`. There is now a note in the working draft to call this out (http://eel.is/c++draft/range.nonprop.cache#note-1).

AFAIK, to get this working in `constexpr` land, you need a wrapper type, for example:

```cpp
struct __deref_tag {};

template<typename _Tp>
struct __cache_wrapper {
    template<typename _Iter>
    constexpr __cache_wrapper(__deref_tag, const _Iter& __i)
    : __t(*__i) {}
    
    _Tp __t;
};

__cache_wrapper<_Ty> __wrapped;
```

and then you can use `_Construct_in_place(__wrapped, __deref_tag{}, _Iter);`.

If you are planning to use `_Non_propagating_cache` for other purposes too, it may make sense to not have `_Emplace_deref` as a member at all, and instead provide the usual `emplace`-style member function template, and then have `join_view` store a `_Non_propagating_cache<__cache_wrapper<...>>`.

Reply:
That is actually even better as I want to use it for split to cache the next subrange